mlir/lib/Conversion/MIGraphXToTosa/MIGraphXToTosa.cpp

```diff
@@ -317,23 +317,51 @@ class DotConverter final : public OpConversionPattern<migraphx::DotOp> {
 
     // check batch dimension. Tosa matmul only allow a single dimension for it,
     // add reshape ops to flatten and restore the original dimension.
-    ArrayRef<int64_t> orgOutDims = outputTy.getShape();
+    SmallVector<int64_t, 5> orgOutDims(outputTy.getShape());
     RankedTensorType newOutType = RankedTensorType::get(orgOutDims, elementTy);
     size_t outRank = orgOutDims.size();
-
-    if (outRank != 3) { // A, B, Out have the same rank. rank=2 assumes batch=1
-      ArrayRef<int64_t> orgDimsA = in_A.getType().cast<ShapedType>().getShape();
-      ArrayRef<int64_t> orgDimsB = in_B.getType().cast<ShapedType>().getShape();
-      int64_t batchSize = 1;
+    SmallVector<int64_t, 5> orgDimsA(
+        in_A.getType().cast<ShapedType>().getShape());
+    SmallVector<int64_t, 5> orgDimsB(
+        in_B.getType().cast<ShapedType>().getShape());
+    size_t rankA = orgDimsA.size();
+    size_t rankB = orgDimsB.size();
+
+    // A, B, Out have the same rank. rank=2 assumes batch=1.
+    // Here handling special cases.
+    if (outRank != 3 || rankA != rankB ||
+        (outRank == 3 && orgDimsA != orgDimsB)) {
+      int64_t batchSizeA = 1, batchSizeB = 1, batchSizeC = 1;
       for (size_t i = 0; i < outRank - 2; i++) {
-        batchSize *= orgOutDims[i];
+        batchSizeC *= orgOutDims[i];
+      }
+      for (size_t i = 0; i < rankA - 2; i++) {
+        batchSizeA *= orgDimsA[i];
+      }
+      for (size_t i = 0; i < rankB - 2; i++) {
+        batchSizeB *= orgDimsB[i];
       }
-      int64_t newDimsA[3] = {batchSize, orgDimsA[outRank - 2],
+
+      int64_t newDimsA[3] = {batchSizeA, orgDimsA[outRank - 2],
                              orgDimsA[outRank - 1]};
-      int64_t newDimsB[3] = {batchSize, orgDimsB[outRank - 2],
+      int64_t newDimsB[3] = {batchSizeB, orgDimsB[outRank - 2],
                              orgDimsB[outRank - 1]};
-      int64_t newDimsOut[3] = {batchSize, orgOutDims[outRank - 2],
+      int64_t newDimsOut[3] = {batchSizeC, orgOutDims[outRank - 2],
                                orgOutDims[outRank - 1]};
+      if (batchSizeA != batchSizeB) {
```

Comment:
Would it possible add a test to check this case ?
(If I am not mistaken, the existing test checks for same batchsize)

Reply:
ah ignore... I just saw multibroadcast doing that.